presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaFilterManager.java

```diff
@@ -0,0 +1,319 @@
+                }
+                else {
+                    Marker lowMark = ranges.getSpan().getLow();
+                    low = getLowByLowMark(lowMark).orElse(low);
+                    Marker highMark = ranges.getSpan().getHigh();
+                    high = getHighByHighMark(highMark).orElse(high);
+                }
+            }
+        }
+        if (high != INVALID_KAFKA_RANGE_INDEX) {
+            high = high + 1;
+        }
+        return Optional.of(new Range(low, high));
+    }
+
+    private static Set<Long> getSingleSet(Domain domain, Set<Long> sourceValues)
+    {
+        requireNonNull(sourceValues, "sourceValues is none");
+        if (domain.isSingleValue()) {
+            long singleValue = (long) domain.getSingleValue();
+            return sourceValues.stream().filter(sourceValue -> sourceValue == singleValue).collect(toImmutableSet());
+        }
+        else {
+            ValueSet valueSet = domain.getValues();
+            if (valueSet instanceof SortedRangeSet) {
+                Ranges ranges = ((SortedRangeSet) valueSet).getRanges();
+                List<io.prestosql.spi.predicate.Range> rangeList = ranges.getOrderedRanges();
+                if (rangeList.stream().allMatch(io.prestosql.spi.predicate.Range::isSingleValue)) {
+                    return rangeList.stream()
+                            .map(range -> (Long) range.getSingleValue())
+                            .filter(sourceValues::contains)
+                            .collect(toImmutableSet());
+                }
+                else {
+                    long low = 0;
+                    long high = Long.MAX_VALUE;
+
+                    Marker lowMark = ranges.getSpan().getLow();
+                    low = maxLow(low, lowMark);
+                    Marker highMark = ranges.getSpan().getHigh();
+                    high = minHigh(high, highMark);
+                    final long finalLow = low;
+                    final long finalHigh = high;
+                    return sourceValues.stream()
+                            .filter(item -> item >= finalLow && item <= finalHigh)
+                            .collect(toImmutableSet());
+                }
+            }
+        }
+        return sourceValues;
+    }
+
+    private static long minHigh(long high, Marker highMark)
+    {
+        Optional<Long> highByHighMark = getHighByHighMark(highMark);
+        if (highByHighMark.isPresent()) {
+            high = Long.min(highByHighMark.get(), high);
+        }
+        return high;
+    }
+
+    private static long maxLow(long low, Marker lowMark)
+    {
+        Optional<Long> lowByLowMark = getLowByLowMark(lowMark);
+        if (lowByLowMark.isPresent()) {
+            low = Long.max(lowByLowMark.get(), low);
+        }
+        return low;
+    }
+
+    private static Optional<Long> getHighByHighMark(Marker highMark)
+    {
+        if (!highMark.isUpperUnbounded() && highMark.getValueBlock().isPresent()) {
```

Comment:
It feels to me that if `!highMark.isUpperUnbounded()` then `highMark.getValueBlock().isPresent()` should always be true. Can we make it `checkArgument` instead?